### DISCUSSION TITLE
BottomSheet v5

### DISCUSSION THREAD

**Discussion by gorhom** [gorhom/react-native-bottom-sheet#1125] [Ideas]
## Overview

This release will be mainly focus on rewriting the current gesture apis with **GH v2** and provide Web support. However, there are other areas that we could improve such accessibility, performance and improve keyboard handling on Android but considered as a non-release blocker.

## Goals
- [x] [BLOCKER] Introduce Web support (#1150). 
- [x] [BLOCKER] Rewrite the package with Gesture Handler v2 api (#1126).
- [x] [BLOCKER] Ensure the package compatibility with Reanimated v3 (#1324).
- [ ] Provide testing mocks to all exports.
- [x] Improve accessibility support (6203c18acc9f8dc3a31af5bf5ad80e368deceb52).
- [ ] Optimise performance.
- [x] Improve keyboard handling (a1ec74dbbc85476bb39f3637e9a97214e0cad9a0).

## Testing

```bash
yarn add ssh://git@github.com:gorhom/react-native-bottom-sheet#v5
```

## Support

Any help will be welcome, from testing, ideation or contributing <3

**Comment by iamvucms** [comment]
Did you support multiple BottomSheet on current version ?

**Reply by gorhom** [reply]
current version does support multiple bottom sheets as modal stack! check the examples

**Comment by gorhom** [comment]
here is a PR that will introduce support for Web

https://github.com/gorhom/react-native-bottom-sheet/pull/1150

**Comment by pierroo** [comment]
Where can we see the actual change this new version provides?
Since I had to update reanimated to v3 seems like I have no other choice but use this new bottom sheet v5 version, but I'm afraid to use it in production for it is only in alpha.
And I can't find any documentations about new stuff and point that we should be careful about etc.
@gorhom please let me know? :)

On the other hand, in a bottom sheet, when I use BottomSheetScrollView the scrolling isn't always working with reanimated v3 and bottomsheet v4.
It seems rather random, sometimes it scrolls down as expected within the content, sometimes the scroll within doesnt work at all and will only work when swiping to leave the bottomsheet.
So I feel like it might be due to incompatibility between reanimated v3 and bottomsheet v4 but I have no way of knowing.

**Comment by lucksp** [comment]
any chance the new version will include an updated package to `react-native-gesture-handler` so it's not on v1, but v2 instead?  I am wondering if this is why my android compilations are failing due to this dependency being `defined multiple times`?  Just a thought.

**Comment by LucasReinaldo** [comment]
hey @gorhom 👋 any chance of bumping react native reanimated to v3 in the current v4 version?

**Reply by spsaucier** [reply]
For us, bottom-sheet's reanimated v2 dependency is the sole remaining is a blocker to upgrading React Native.

**Reply by jiroscripts** [reply]
Yes, I can't upgrade to RN 0.72 either.
See: https://github.com/software-mansion/react-native-reanimated/issues/4522

**Reply by eithe** [reply]
@thibaultcapelli If you are willing to use the BottomSheet v5 alpha release (which this topic is about, and that I have been using in production for some time now), you can update your package.json as such (this is what I'm running, but I noticed there are some version increases to gesture handler (2.14.0) and reanimated (3.6.0) that I haven't updated to yet):

```
"dependencies": {
  ...
  "@gorhom/bottom-sheet": "https://github.com/gorhom/react-native-bottom-sheet#v5",
  "react-native-gesture-handler": "2.13.4",
  "react-native-reanimated": "3.5.4",
  ...
}
```

**Reply by jiroscripts** [reply]
Hello @eithe 

I am currently trying this v5 in development mode. The documentation doesn't mention any v5 in the readme and there is nothing clear about it. Thanks for your feedback as I was wondering if it was stable enough for production.

**Reply by eithe** [reply]
No worries @thibaultcapelli. We are not very advanced users of BottomSheet, but v5 works for what we use it for in our app (I wrote about it [here](https://github.com/gorhom/react-native-bottom-sheet/issues/912#issuecomment-1413400443) so been in use/production for almost a year)

**Reply by spsaucier** [reply]
@eithe How are you building it? When I include v5 it doesn't have a `lib` folder, and I can't build it from that source, either (missing tsconfig, at least).

**Reply by eithe** [reply]
@spsaucier Hi. Sorry, not sure I fully understand your question or what you mean by a `lib` folder? I'm not doing anything special other than updating my package.json to point to the v5 version (+ whatever was needed in the app to go to a newer gesture handler + reanimated), there is no such folder involved (edit: to use v5)?

**Reply by spsaucier** [reply]
@eithe 
v4 in node_modules after installing includes a `lib` folder:
<img width="139" alt="Screenshot 2023-11-30 at 10 18 56" src="https://github.com/gorhom/react-native-bottom-sheet/assets/596143/b981d4ee-1b72-4031-b9bc-29b9ef7991fd">

v5 does not:
<img width="150" alt="Screenshot 2023-11-30 at 10 18 32" src="https://github.com/gorhom/react-native-bottom-sheet/assets/596143/42cf839f-32a2-4014-95f8-3164320021c7">
Its `package.json` says that the lib folder is where the entry point is:
<img width="281" alt="Screenshot 2023-11-30 at 10 18 15" src="https://github.com/gorhom/react-native-bottom-sheet/assets/596143/81779dd8-a312-4669-9c86-63572c79a2c1">
And since there is none, the app can't find the package.
<img width="652" alt="Screenshot 2023-11-30 at 10 17 30" src="https://github.com/gorhom/react-native-bottom-sheet/assets/596143/51abd82f-d047-4948-9171-32b8d8797682">

**Reply by jiroscripts** [reply]
@spsaucier I can confirm there I have no lib folder in v5 and I have also the typescript issue but the package still works

**Reply by spsaucier** [reply]
Oh, thank you @thibaultcapelli. That is very strange, but I can confirm - v5 is working even with the TS issue.

**Reply by spsaucier** [reply]
While it compiles for the application, it does fail for jest tests. I'll have to figure out how to make them work with the uncompiled version.

**Reply by jiroscripts** [reply]
@spsaucier Just in case, do not follow this command `ssh://git@github.com:gorhom/react-native-bottom-sheet#v5` if you want the typescript definitions. Instead you can do `@gorhom/bottom-sheet@alpha` to install the latest alpha release (currently 5.0.0-alpha.4)

**Comment by markymc** [comment]
@gorhom Thank you very much for getting started on web support. I've been trying to make use of it but am facing some troubles. I'm wondering if you have any plans to rebase the v5 branch onto the latest v4 release? I had a quick look but I got a little out of my depth with conflicts. 🙇‍♂️

**Comment by jiroscripts** [comment]
@gorhom Could you please update the install command in your post to `yarn add @gorhom/bottom-sheet@alpha` ?

Cause the typescript definitions are not builded in the code repository but only in the releases.

**Comment by spsaucier** [comment]
@gorhom Will there be an alpha release to include `enableDynamicSizing` soon? As it stands, we are stuck on v4 until then, unless we rewrite our implementations back to the old style.

**Reply by gorhom** [reply]
im wrapping it up at the moment.. expect an alpha release today or tomorrow

**Comment by imranbarbhuiya** [comment]
Is there a migration guide/change log available for v5 alpha upgrade?

**Comment by dmregister** [comment]
@gorhom when do you expect v5 to out of alpha and into RC?

**Comment by gorhom** [comment]
Hi everyone, it is been a while 😅

before cutting the RC for v5, i wanted to refactor and improve the snapping sequencing which i have been working on for a while, here is a PR that should introduce this blocker improvement and i would like from you to test it if possible.

https://github.com/gorhom/react-native-bottom-sheet/pull/1784

**Comment by kevinschaffter** [comment]
Just started using this new version and it is fantastic, thank you 🥳 

One small issue I noticed is that I needed to explicitly pass in `enabledDynamicSizing={false}` even though in the docs it says it's false by default. This became especially apparent when my sheet had content that went past my largest snap point.

**Reply by gorhom** [reply]
Thanks @kevinschaffter for testing v5, dynamic sizing is enabled by default for this version. And yes it will fill up the parent if you have content that is large, that's why I introduced another property to limit that sizing 'maxDynamicContentSize'

**Reply by kevinschaffter** [reply]
Ah ok, sounds good. I was just confused at first as the docs for V5 said it was disabled by default.

**Comment by eithe** [comment]
Also tested this version now (for anyone else wondering I added/changed my package.json to say `"@gorhom/bottom-sheet": "gorhom/react-native-bottom-sheet#refactor/refactor-snap-points-reactions"`, but note that you will miss typescript types when you do that).

Things seem to work as expected with this version which is great since I have been stuck on 5.0.0-alpha.4 (couldn't actually get my bottomsheets to appear at all on alpha.5 and .6 (haven't tried later versions, and didn't spend much time to debug)).

Like @kevinschaffter I also had to set `enabledDynamicSizing={false}` or else I got an error on that ("Invariant Violation: 'enableDynamicSizing' is enabled but children type is not integrated with the library ! expected children types are: ...")

Thanks for the update @gorhom, looks like you are on the right track with this :)

**Reply by gorhom** [reply]
Thanks for testing, I'll be merging the pr this weekend and prepare the first RC.

**Reply by gegana** [reply]
@gorhom would it be possible to change the default to false? Currently I would have to patch the package to avoid editing 30+ files :)

**Comment by khushal87** [comment]
Hey, @gorhom, how is it going while testing the package with a new architecture that uses the latest version of react-native and react-native-reanimated? When can we expect a stable release that would support the same?

**Comment by akylbek-hostaways** [comment]
Hi @gorhom. It is possible to add testID for this release?
